internal/packer_registry/types.iterations.go

```diff
@@ -0,0 +1,67 @@
+package packer_registry
+
+import (
+	"fmt"
+	"os"
+
+	git "github.com/go-git/go-git/v5"
+	// "github.com/hashicorp/hcp-sdk-go/clients/cloud-packer-service/preview/2021-04-30/models"
+)
+
+type Iteration struct {
+	ID           string
+	AncestorSlug string
+	Fingerprint  string
+	RunUUID      string
+	Labels       map[string]string
+	builds       Builds
+}
+
+type IterationOptions struct {
+	TemplateBaseDir string
+}
+
+func GetGitFingerprint(opts IterationOptions) (string, error) {
+	r, err := git.PlainOpenWithOptions(opts.TemplateBaseDir, &git.PlainOpenOptions{
+		DetectDotGit: true,
+	})
+	if err != nil {
+		return "", fmt.Errorf("Packer was unable to load a git sha. "+
+			"If your Packer template is not in a git repo, please add a unique "+
+			"template fingerprint using the env var HCP_PACKER_BUILD_FINGERPRINT. "+
+			"Error: %s", err)
+	}
+	// The config can be used to retrieve user identity. for example,
+	// c.User.Email. Leaving in but commented because I'm not sure we care
+	// about this identity right now. - Megan
+	//
+	// c, err := r.ConfigScoped(config.GlobalScope)
+	// if err != nil {
+	// 	return "", fmt.Errorf("Error setting git scope", err)
+	// }
+	ref, _ := r.Head()
+	// log.Printf("Author: %v, Commit: %v\n", c.User.Email, ref.Hash())
+	return ref.Hash().String(), nil
+}
+
+// NewIteration returns a pointer to an Iteration that can be used for storing Packer build details needed by PAR.
+func NewIteration(opts IterationOptions) (*Iteration, error) {
+	i := Iteration{
+		builds: NewBuilds(),
+	}
+
+	// By default we try to load a Fingerprint from the environment variable.
+	// If no variable is defined we should try to load a fingerprint from Git, or other VCS.
+	i.Fingerprint = os.Getenv("HCP_PACKER_BUILD_FINGERPRINT")
+
+	// Get a git sha.
+	if i.Fingerprint == "" {
+		fp, err := GetGitFingerprint(opts)
+		if err != nil {
+			return nil, err
+		}
+		i.Fingerprint = fp
+	}
```

Comment:
I would make of that fingerprint a HCL parameter, and fail if it is not set.

```
something {
  fingerprint = git-sha(path.cwd)
}
or
something {
  fingerprint = dirhash(path.cwd)
}
or 
something {
  fingerprint = "42"
}
```

Reply:
`hash` could be a native packer function or something in a plugin. I have a slight preference for a plugin, because it makes it easier to udpate/fix/release.